RecoLocalCalo/EcalRecAlgos/src/EcalRecHitBuilderKernels.cu

```diff
@@ -0,0 +1,738 @@
+      kFixedG6,
+      kFixedG1,
+      kFixedG0,
+      kNonRespondingIsolated,
+      kDeadVFE,
+      kDeadFE,
+      kNoDataNoTP      
+    };
+    
+    
+    
+    
+    
+    __global__
+    void kernel_create_ecal_rehit(
+      // configuration 
+      int const* ChannelStatusToBeExcluded,
+      uint32_t ChannelStatusToBeExcludedSize,   
+      bool const killDeadChannels,
+      bool const recoverEBIsolatedChannels,
+      bool const recoverEEIsolatedChannels,
+      bool const recoverEBVFE,             
+      bool const recoverEEVFE,             
+      bool const recoverEBFE,             
+      bool const recoverEEFE,              
+      float const EBLaserMIN,
+      float const EELaserMIN,
+      float const EBLaserMAX,
+      float const EELaserMAX,
+      // for flags setting
+      int const* expanded_v_DB_reco_flags,    // FIXME AM: to be checked
+      uint32_t const* expanded_Sizes_v_DB_reco_flags,
+      uint32_t const* expanded_flagbit_v_DB_reco_flags,
+      uint32_t expanded_v_DB_reco_flagsSize,
+      uint32_t flagmask,
+      // conditions
+      float const* adc2gev,
+      float const* intercalib,
+      uint16_t const* status,
+      float const* apdpnrefs,
+      float const* alphas,
+      // input for transparency corrections
+      float const* p1,
+      float const* p2,
+      float const* p3,
+      edm::TimeValue_t const* t1,
+      edm::TimeValue_t const* t2,
+      edm::TimeValue_t const* t3,  
+      // input for linear corrections
+      float const* lp1,
+      float const* lp2,
+      float const* lp3,
+      edm::TimeValue_t const* lt1,
+      edm::TimeValue_t const* lt2,
+      edm::TimeValue_t const* lt3,                    
+      // time, used for time dependent corrections
+      edm::TimeValue_t const event_time,
+      // input
+      uint32_t const* did_eb,
+      uint32_t const* did_ee,
+      ::ecal::reco::StorageScalarType const* amplitude_eb,   // in adc counts  
+      ::ecal::reco::StorageScalarType const* amplitude_ee,   // in adc counts  
+      ::ecal::reco::StorageScalarType const* time_eb,   
+      ::ecal::reco::StorageScalarType const* time_ee,   
+      ::ecal::reco::StorageScalarType const* chi2_eb,   
+      ::ecal::reco::StorageScalarType const* chi2_ee,   
+      uint32_t const* flags_eb,
+      uint32_t const* flags_ee,
+      // output
+      uint32_t *did,
+      ::ecal::reco::StorageScalarType* energy,   // in energy [GeV]  
+      ::ecal::reco::StorageScalarType* time,  
+      ::ecal::reco::StorageScalarType* chi2,  
+      uint32_t* flagBits,
+      uint32_t* extra,
+      // other
+      int const nchannels,
+      uint32_t const offsetForInput,
+      uint32_t const offsetForHashes                     
+    ) {
+      
+      
+      //       
+      //    NB: energy   "type_wrapper<reco::StorageScalarType, L>::type" most likely std::vector<float>
+      //       
+      
+      int ch = threadIdx.x + blockDim.x*blockIdx.x;
+      
+      if (ch < nchannels) {
```

Comment:
In general we try to make the code support arbitrary "grid" size, even running with a single block and thread.
This requires changing this to
```suggestion
      for (int ch = threadIdx.x + blockDim.x*blockIdx.x; ch < nchannels; ch += blockDim.x*gridDim.x) {
```

do you see any problems in using this approach here ?

Reply:
(the lambba suggested above should be defined outside the for loop)

Reply:
Inside the kernel we were targeting 1 single channel at a time ... if we do the for loop we run more than one channel?

I think already now the grid size is arbitrary ... or maybe I am missing something ...

Reply:
faster to discuss in chat :-)

Reply:
Ok, changed.
It works properly.